File: python/double_pendulum/simulation.py
"""Fixed-step driver for one (or two) double pendulums.

Mirrors the TypeScript ``Simulation`` class: seeded/reproducible forcing,
damping, a moving pivot, user impulses, a chaos-comparison twin, and a capped
history buffer for charts and export.
"""

from __future__ import annotations

import math
from typing import Callable, List, Optional, Tuple

from .angles import state_separation
from .equations import (
    damping_power,
    derivative,
    energy,
    external_power,
    pivot_acceleration,
)
from .forcing import deterministic_torque, random_torque
from .integrator import rk4_step
from .seeded_random import mulberry32
from .types import HistoryRecord, Params, State

MAX_HISTORY = 4000      # cap retained samples to protect memory
IMPULSE_DURATION = 0.12  # seconds a torque impulse is applied


class UnstableError(Exception):
    """Raised when the integration blows up to NaN/Inf."""


class Simulation:
    def __init__(self, params: Params) -> None:
        self.params: Params = params
        self.x: State = (0.0, 0.0, 0.0, 0.0)
        self.x2: Optional[State] = None
        self.t: float = 0.0
        self.history: List[HistoryRecord] = []
        self.last_torque: Tuple[float, float] = (0.0, 0.0)

        self._rng: Callable[[], float] = mulberry32(1)
        self._impulse_value: List[float] = [0.0, 0.0]
        self._impulse_left: List[float] = [0.0, 0.0]
        self._d0: float = 1e-9

        self.reset(params)

    def reset(self, params: Optional[Params] = None) -> None:
        if params is not None:
            self.params = params
        p = self.params
        self.x = (p.theta1_0, p.theta2_0, p.omega1_0, p.omega2_0)
        self.t = 0.0
        self.history = []
        self._rng = mulberry32(p.seed)
        self._impulse_value = [0.0, 0.0]
        self._impulse_left = [0.0, 0.0]
        if p.comparison:
            self.x2 = (p.theta1_0 + p.epsilon, p.theta2_0, p.omega1_0, p.omega2_0)
            self._d0 = state_separation(self.x, self.x2) or 1e-9
        else:
            self.x2 = None
        self._record_sample((0.0, 0.0))

    def set_comparison(self, on: bool, epsilon: float) -> None:
        self.params = self.params.copy()
        self.params.comparison = on
        self.params.epsilon = epsilon
        if on:
            self.x2 = (self.x[0] + epsilon, self.x[1], self.x[2], self.x[3])
            self._d0 = state_separation(self.x, self.x2) or 1e-9
        else:
            self.x2 = None

    def apply_impulse(self, joint: int, sign: int) -> None:
        self._impulse_value[joint] = sign * self.params.impulse_magnitude
        self._impulse_left[joint] = IMPULSE_DURATION

    def step(self) -> None:
        """Advance the simulation by one fixed physics step."""

        p = self.params
        dt = p.dt
        t0 = self.t

        # Seeded random samples are drawn once per step and held constant across
        # the four RK4 stages, guaranteeing reproducibility and identical forcing
        # for both pendulums in comparison mode.
        rt1 = random_torque(p.forcing1, self._rng())
        rt2 = random_torque(p.forcing2, self._rng())
        imp0 = self._impulse_value[0] if self._impulse_left[0] > 0 else 0.0
        imp1 = self._impulse_value[1] if self._impulse_left[1] > 0 else 0.0

        def torque_at(t: float) -> Tuple[float, float]:
            return (
                deterministic_torque(p.forcing1, t) + rt1 + imp0,
                deterministic_torque(p.forcing2, t) + rt2 + imp1,
            )

        def f(x: State, t: float) -> State:
            return derivative(x, p, torque_at(t), pivot_acceleration(p, t))

        self.x = rk4_step(self.x, t0, dt, f)
        if self.x2 is not None:
            self.x2 = rk4_step(self.x2, t0, dt, f)
        self.t = t0 + dt

        self._impulse_left[0] = max(0.0, self._impulse_left[0] - dt)
        self._impulse_left[1] = max(0.0, self._impulse_left[1] - dt)

        tau = torque_at(t0)
        self.last_torque = tau

        if not self._is_finite(self.x) or (
            self.x2 is not None and not self._is_finite(self.x2)
        ):
            raise UnstableError(
                "Simulation became numerically unstable (NaN/Inf). Reduce the "
                "time step or forcing amplitude, then reset."
            )
        self._record_sample(tau)

    @staticmethod
    def _is_finite(x: State) -> bool:
        return all(math.isfinite(v) for v in x)

    def _record_sample(self, tau: Tuple[float, float]) -> None:
        p = self.params
        e = energy(self.x, p)
        pext = external_power(tau, self.x)
        pdamp = damping_power(self.x, p)

        d: Optional[float] = None
        lyap: Optional[float] = None
        if self.x2 is not None:
            d = state_separation(self.x, self.x2)
            if self.t > 0 and d > 0:
                lyap = (1 / self.t) * math.log(d / self._d0)

        self.history.append(
            HistoryRecord(
                t=self.t,
                th1=self.x[0],
                th2=self.x[1],
                w1=self.x[2],
                w2=self.x[3],
                ke=e["ke"],
                pe=e["pe"],
                e=e["total"],
                tau1=tau[0],
                tau2=tau[1],
                pext=pext,
                pdamp=pdamp,
                d=d,
                lyap=lyap,
            )
        )
        if len(self.history) > MAX_HISTORY:
            self.history.pop(0)

    def latest(self) -> HistoryRecord:
        return self.history[-1]


__all__ = ["Simulation", "UnstableError", "MAX_HISTORY", "IMPULSE_DURATION"]

File: src/physics/simulation.ts
import { Params, State, HistoryRecord } from '../types';
import { rk4Step, Deriv } from './integrator';
import {
  derivative, pivotAcceleration, energy, externalPower, dampingPower,
} from './equations';
import { deterministicTorque, randomTorque } from './forcing';
import { stateSeparation } from './angles';
import { mulberry32 } from '../utils/seededRandom';

const MAX_HISTORY = 4000;      // cap retained samples to protect the browser
const IMPULSE_DURATION = 0.12; // seconds a torque impulse is applied

export class UnstableError extends Error {}

// Drives one (or two, in comparison mode) double pendulums with a fixed-step
// RK4 integrator, seeded/reproducible forcing, damping, moving pivot and user
// impulses, while recording a capped history for charts and export.
export class Simulation {
  params: Params;
  x: State = [0, 0, 0, 0];
  x2: State | null = null;
  t = 0;
  history: HistoryRecord[] = [];
  lastTorque: [number, number] = [0, 0];

  private rng: () => number = mulberry32(1);
  private impulseValue: [number, number] = [0, 0];
  private impulseLeft: [number, number] = [0, 0];
  private d0 = 1e-9;

  constructor(params: Params) {
    this.params = params;
    this.reset(params);
  }

  reset(params?: Params): void {
    if (params) this.params = params;
    const p = this.params;
    this.x = [p.theta1_0, p.theta2_0, p.omega1_0, p.omega2_0];
    this.t = 0;
    this.history = [];
    this.rng = mulberry32(p.seed);
    this.impulseValue = [0, 0];
    this.impulseLeft = [0, 0];
    if (p.comparison) {
      this.x2 = [p.theta1_0 + p.epsilon, p.theta2_0, p.omega1_0, p.omega2_0];
      this.d0 = stateSeparation(this.x, this.x2) || 1e-9;
    } else {
      this.x2 = null;
    }
    this.recordSample([0, 0]);
  }

  setComparison(on: boolean, epsilon: number): void {
    this.params = { ...this.params, comparison: on, epsilon };
    if (on) {
      this.x2 = [this.x[0] + epsilon, this.x[1], this.x[2], this.x[3]];
      this.d0 = stateSeparation(this.x, this.x2) || 1e-9;
    } else {
      this.x2 = null;
    }
  }

  applyImpulse(joint: 0 | 1, sign: 1 | -1): void {
    this.impulseValue[joint] = sign * this.params.impulseMagnitude;
    this.impulseLeft[joint] = IMPULSE_DURATION;
  }

  // Advance the simulation by one fixed physics step.
  step(): void {
    const p = this.params;
    const dt = p.dt;
    const t0 = this.t;

    // Seeded random samples are drawn once per step and held constant across
    // the four RK4 stages, guaranteeing reproducibility and identical forcing
    // for both pendulums in comparison mode.
    const rt1 = randomTorque(p.forcing1, this.rng());
    const rt2 = randomTorque(p.forcing2, this.rng());
    const imp0 = this.impulseLeft[0] > 0 ? this.impulseValue[0] : 0;
    const imp1 = this.impulseLeft[1] > 0 ? this.impulseValue[1] : 0;

    const torqueAt = (t: number): [number, number] => [
      deterministicTorque(p.forcing1, t) + rt1 + imp0,
      deterministicTorque(p.forcing2, t) + rt2 + imp1,
    ];

    const f: Deriv = (x, t) =>
      derivative(x, p, torqueAt(t), pivotAcceleration(p, t));

    this.x = rk4Step(this.x, t0, dt, f);
    if (this.x2) this.x2 = rk4Step(this.x2, t0, dt, f);
    this.t = t0 + dt;

    this.impulseLeft[0] = Math.max(0, this.impulseLeft[0] - dt);
    this.impulseLeft[1] = Math.max(0, this.impulseLeft[1] - dt);

    const tau = torqueAt(t0);
    this.lastTorque = tau;

    if (!this.isFinite(this.x) || (this.x2 && !this.isFinite(this.x2))) {
      throw new UnstableError(
        'Simulation became numerically unstable (NaN/Inf). Reduce the time step or forcing amplitude, then reset.'
      );
    }
    this.recordSample(tau);
  }

  private isFinite(x: State): boolean {
    return x.every((v) => Number.isFinite(v));
  }

  private recordSample(tau: [number, number]): void {
    const p = this.params;
    const { ke, pe, total } = energy(this.x, p);
    const pext = externalPower(tau, this.x);
    const pdamp = dampingPower(this.x, p);

    let d: number | null = null;
    let lyap: number | null = null;
    if (this.x2) {
      d = stateSeparation(this.x, this.x2);
      if (this.t > 0 && d > 0) {
        lyap = (1 / this.t) * Math.log(d / this.d0);
      }
    }

    this.history.push({
      t: this.t,
      th1: this.x[0], th2: this.x[1], w1: this.x[2], w2: this.x[3],
      ke, pe, e: total,
      tau1: tau[0], tau2: tau[1],
      pext, pdamp,
      d, lyap,
    });
    if (this.history.length > MAX_HISTORY) this.history.shift();
  }

  latest(): HistoryRecord {
    return this.history[this.history.length - 1];
  }
}
